api/plugin.go

```diff
@@ -0,0 +1,602 @@
+	"github.com/gotify/server/auth"
+	"github.com/gotify/server/model"
+	"github.com/gotify/server/plugins/holder"
+	"github.com/gotify/server/plugins/loader"
+
+	papiv1 "github.com/gotify/plugin-api"
+)
+
+// The PluginDatabase interface for encapsulating database access.
+type PluginDatabase interface {
+	GetUsers() []*model.User
+	GetUserByID(id uint) *model.User
+	CreateApplication(application *model.Application) error
+	GetApplicationByToken(token string) *model.Application
+	GetPluginConfByUser(userid uint) []*model.PluginConf
+	GetPluginConfByUserAndPath(userid uint, path string) *model.PluginConf
+	CreatePluginConf(p *model.PluginConf) error
+	GetPluginConfByToken(token string) *model.PluginConf
+	UpdatePluginConf(p *model.PluginConf) error
+	CreateMessage(message *model.Message) error
+}
+
+// The PluginAPI provides handlers for managing plugins.
+type PluginAPI struct {
+	mutex          sync.Mutex
+	Notifier       Notifier
+	DB             PluginDatabase
+	Holder         holder.PluginInstanceHolder
+	CustomRouteMux *gin.RouterGroup
+}
+
+// GetPlugins returns all plugins a user has.
+// swagger:operation GET /plugin plugin getPlugins
+//
+// Return all plugins.
+//
+// ---
+// consumes: [application/json]
+// produces: [application/json]
+// security: [clientTokenHeader: [], clientTokenQuery: [], basicAuth: []]
+// responses:
+//   200:
+//     description: Ok
+//     schema:
+//       type: array
+//       items:
+//         $ref: "#/definitions/PluginConf"
+//   401:
+//     description: Unauthorized
+//     schema:
+//         $ref: "#/definitions/Error"
+//   500:
+//     description: Internal Server Error
+//     schema:
+//         $ref: "#/definitions/Error"
+func (c *PluginAPI) GetPlugins(ctx *gin.Context) {
+	userID := auth.GetUserID(ctx)
+	plugins := c.DB.GetPluginConfByUser(userID)
+	result := make([]model.PluginConfExternal, 0)
+	for _, conf := range plugins {
+		if inst, err := c.Holder.GetInstance(conf.Token); err == nil {
+			p, err := c.Holder.PluginHolder.Get(conf.ModulePath)
+			name := "[UNNAMED]"
+			if err == nil && p.GetPluginInfo().Name != "" {
+				name = p.GetPluginInfo().Name
+			}
+			capabilities := make([]string, 0)
+			if _, ok := inst.(papiv1.Plugin); ok {
+				capabilities = append(capabilities, "v1")
+				if _, ok := inst.(papiv1.Configurer); ok {
+					capabilities = append(capabilities, "v1.configurer")
+				}
+				if _, ok := inst.(papiv1.Displayer); ok {
+					capabilities = append(capabilities, "v1.displayer")
+				}
+				if _, ok := inst.(papiv1.Messenger); ok {
+					capabilities = append(capabilities, "v1.messeger")
+				}
+				if _, ok := inst.(papiv1.Webhooker); ok {
+					capabilities = append(capabilities, "v1.webhooker")
+				}
+				if _, ok := inst.(papiv1.Storager); ok {
+					capabilities = append(capabilities, "v1.storager")
+				}
+			}
+			result = append(result, model.PluginConfExternal{
+				ID:           conf.ID,
+				Name:         name,
+				Token:        conf.Token,
```

Comment:
To calculate the custom route endpoint.

Reply:
Do we need that? the plugins will probably add an own path to it, so it will be different for every plugin?

Reply:
I'd still prefer using the id as identifier in the url and then also have the PluginID in the url for the custom route. This makes it more conistent with the other apis as they all use integer ids.

> user need to know two pieces of information to calculate the endpoint

I guess not? The plugin can set custom paths so it is not really useful to do this. The plugin should output the full path via the Displayer interface. Only the plugin will know on which path it has registered the webhook url.

> two authentication methods would coexist in /plugin/:id/*

I don't see this as a problem. Things like this do also already exist `POST /message` requires an authentication token and `GET /message` a client token. 
The custom plugin path could also be served on an other path something like `domain.com/custom/plugin/:plugintoken` but I'd prefer not to because it belongs to `/plugin`.